source/extensions/injected_credentials/common/credential.h

```diff
@@ -0,0 +1,73 @@
+#pragma once
+
+#include "envoy/common/exception.h"
+#include "envoy/http/header_map.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace InjectedCredentials {
+namespace Common {
+
+/**
+ * Credential injector injects credential to HTTP headers.
+ */
+class CredentialInjector {
+public:
+  virtual ~CredentialInjector() = default;
+
+  /*
+   * Request for a credential from a given source.
+   */
+  class Request {
+  public:
+    virtual ~Request() = default;
+
+    /**
+     * Signals that request should be cancelled.
+     */
+    virtual void cancel() PURE;
+  };
+  using RequestPtr = std::unique_ptr<Request>;
+
+  /**
+   * Notifies caller of the request status.
+   */
+  class Callbacks {
+  public:
+    virtual ~Callbacks() = default;
+
+    /**
+     * Called when the async request for credentials succeeds.
+     */
+    virtual void onSuccess() PURE;
+
+    /**
+     * Called when the async request for credentials fails.
+     * @param reason supplies the failure reason.
+     */
+    virtual void onFailure(const std::string& reason) PURE;
```

Comment:
personal perference: `absl::string_view` would be better than `const std::string&`